rfc/016-node-access.md

```diff
@@ -0,0 +1,176 @@
+# Node Access
+
+## Background
+
+A production Constellation cluster is currently configured not to allow any kind of remote administrative access.
+This choice is deliberate: any mechanism for remote accesss can potentially be exploited, or may leak sensitive data.
+
+However, some operations on a Kubernetes cluster require some form of access to the nodes.
+A good class of examples are etcd cluster maintenance tasks, like backup and recovery, or emergency operations like removing a permanently failed member.
+Some kubeadm operations, like certificate rotation, also require some form of cluster access.
+
+While some tasks can be accomplished by DaemonSets, CronJobs and the like, relying on Kubernetes objects is insufficient.
+Executing commands in a Kubernetes pod may fail because Kubernetes is not healthy, etcd is bricked or the network is down.
+Administrative access to the nodes through a side channel would greatly help remediate, or at least debug, those situations.
+
+## Requirements
+
+Constellation admins can log into Constellation nodes for maintenance, subject to the following restrictions:
+
+* Access must be encrypted end-to-end to protect from CSP snooping.
+* Access must be possible even if the Kubernetes API server is down.
+
+Nice-to-have:
+
+* The method of access should not require long-term storage of a second secret.
+* The method of access should be time-limited.
+
+## Proposed Design
+
+Core to the proposal is [certificate-based authentication for OpenSSH](https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Certificate-based_Authentication).
+We can derive a valid SSH key from the Constellation master secret.
+An OpenSSH server on the node accepts certificates issued by this CA key.
+Admins can derive the CA key from the master secret on demand, and issue certificates for arbitrary public keys.
+An example program is in the [Appendix](#appendix).
+
+### Key Details
+
+We use an HKDF to derive an ed25519 private key from the master secret.
+This private key acts as an SSH certificate authority, whose signed certs allow access to cluster nodes.
+Since the master secret is available to both the cluster owner and the nodes, no communication with the cluster is needed to mint valid certificates.
+The choice of curve allows to directly use the derived secret bytes as key.
+This makes the implementation deterministic, and thus the CA key recoverable.
+
+### Server-side Details
+
+An OpenSSH server is added to the node image software stack.
+It's configured with a `TrustedUserCAKeys` file that is empty on startup.
+After initialization, the bootstrapper fills that file with the derived CA's public key.
```

Comment:
Just as a side note: Worker nodes don't have access to the master secret, so we will need to generate and distribute the CA during the join process.

Reply:
Proposing to only keep the CA in the join server, ptal